--- a/net/minecraft/world/level/block/state/BlockState.java
+++ b/net/minecraft/world/level/block/state/BlockState.java
@@ -25,6 +_,60 @@
         super(owner, values, propertiesCodec);
     }
 
+    // Fiddle start - Cache block and item values
+    /**
+     * The {@linkplain net.minecraft.core.Registry#getId index} at which this block state is registered in
+     * {@link Block#BLOCK_STATE_REGISTRY}.
+     *
+     * <p>
+     * This will be -1 until initialized in
+     * {@link org.fiddlemc.fiddle.impl.minecraft.registries.BlockStateRegistry#add}.
+     * </p>
+     */
+    public int indexInBlockStateRegistry = -1;
+
+    /**
+     * The {@linkplain net.minecraft.core.Registry#getId index} at which this block state is registered in
+     * {@link org.fiddlemc.fiddle.impl.minecraft.registries.VanillaOnlyBlockStateRegistry}.
+     *
+     * <p>
+     * This will be -1 until initialized in
+     * {@link org.fiddlemc.fiddle.impl.minecraft.registries.VanillaOnlyBlockStateRegistry#add}.
+     * </p>
+     */
+    public int indexInVanillaOnlyBlockStateRegistry = -1;
+    // Fiddle end - Cache block and item values
+
+    // Fiddle start - Vanilla flag
+    private boolean isVanilla;
+
+    /**
+     * Returns whether this block state is present in vanilla.
+     *
+     * <p>
+     * This does not take into account whether any details of this block state, such as its collision box,
+     * are different (were modified) from vanilla.
+     * </p>
+     *
+     * @return Whether this block state is present in vanilla.
+     */
+    public boolean isVanilla() {
+        return this.isVanilla;
+    }
+
+    /**
+     * Sets {@link #isVanilla} to true.
+     *
+     * <p>
+     * This must only be called at most once,
+     * before the block state is registered to {@link Block#BLOCK_STATE_REGISTRY}.
+     * </p>
+     */
+    public void markAsVanilla() {
+        this.isVanilla = true;
+    }
+    // Fiddle end - Vanilla flag
+
     @Override
     protected BlockState asState() {
         return this;
