--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -196,12 +_,27 @@
                 } else {
                     Holder<Item> holder = Item.STREAM_CODEC.decode(buffer);
                     DataComponentPatch dataComponentPatch = codec.decode(buffer);
-                    return new ItemStack(holder, varInt, dataComponentPatch);
+                    // Fiddle start - Item mapping - Reverse - Apply at ItemStack decoding
+                    org.fiddlemc.fiddle.impl.packetmapping.item.reverse.@Nullable ItemMappingReverser itemMappingReverser = ((org.fiddlemc.fiddle.impl.clientview.ClientViewImpl) buffer.getClientViewOrFallback()).getItemMappingReverser();
+                    ItemStack received = new ItemStack(holder, varInt, dataComponentPatch);
+                    if (itemMappingReverser != null) {
+                        received = itemMappingReverser.reverseMappingIfPossible(received);
+                    }
+                    return received;
+                    // Fiddle end - Item mapping - Reverse - Apply at ItemStack decoding
                 }
             }
 
             @Override
             public void encode(RegistryFriendlyByteBuf buffer, ItemStack value) {
+                // Fiddle start - Item mapping - Implementation - Apply at ItemStack encoding
+                org.fiddlemc.fiddle.impl.packetmapping.item.ItemMappingContextImpl itemMappingContext = new org.fiddlemc.fiddle.impl.packetmapping.item.ItemMappingContextImpl(buffer.getClientViewOrFallback(), buffer.isEncodingItemFrameItem, buffer.isEncodingStonecutterRecipeResult);
+                ItemStack mappedValue = org.fiddlemc.fiddle.impl.packetmapping.item.ItemMappingPipelineImpl.get().apply(value, itemMappingContext);
+                this.encodeUnmapped(buffer, mappedValue);
+            }
+
+            private void encodeUnmapped(RegistryFriendlyByteBuf buffer, ItemStack value) {
+                // Fiddle end - Item mapping - Implementation - Apply at ItemStack encoding
                 if (value.isEmpty() || value.getItem() == null) { // CraftBukkit - NPE fix itemstack.getItem()
                     buffer.writeVarInt(0);
                 } else {
