--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -196,12 +_,20 @@
                 } else {
                     Holder<Item> holder = Item.STREAM_CODEC.decode(buffer);
                     DataComponentPatch dataComponentPatch = codec.decode(buffer);
-                    return new ItemStack(holder, varInt, dataComponentPatch);
+                    return org.fiddlemc.fiddle.impl.packetmapping.item.encloseserverside.EncloseServerSideItemStack.extractServerSideOrKeepReceived(new ItemStack(holder, varInt, dataComponentPatch)); // Fiddle - Item mapping - Enclose server-side - Apply at ItemStack decoding
                 }
             }
 
             @Override
             public void encode(RegistryFriendlyByteBuf buffer, ItemStack value) {
+                // Fiddle start - Item mapping - Implementation - Apply at ItemStack encoding
+                org.fiddlemc.fiddle.impl.packetmapping.item.ItemMappingContextImpl itemMappingContext = new org.fiddlemc.fiddle.impl.packetmapping.item.ItemMappingContextImpl(buffer.getClientViewOrFallback(), buffer.isEncodingItemFrameItem, buffer.isEncodingStonecutterRecipeResult);
+                ItemStack mappedValue = org.fiddlemc.fiddle.impl.packetmapping.item.ItemMappingPipelineImpl.get().apply(value, itemMappingContext);
+                this.encodeUnmapped(buffer, mappedValue);
+            }
+
+            private void encodeUnmapped(RegistryFriendlyByteBuf buffer, ItemStack value) {
+                // Fiddle end - Item mapping - Implementation - Apply at ItemStack encoding
                 if (value.isEmpty() || value.getItem() == null) { // CraftBukkit - NPE fix itemstack.getItem()
                     buffer.writeVarInt(0);
                 } else {
