--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1282,13 +_,15 @@
             // Paper end
             org.spigotmc.WatchdogThread.hasStarted = true; // Paper
             // Paper start - Add onboarding message for initial server start
-            if (io.papermc.paper.configuration.GlobalConfiguration.isFirstStart) {
-                LOGGER.info("*************************************************************************************");
-                LOGGER.info("This is the first time you're starting this server.");
-                LOGGER.info("It's recommended you read our 'Getting Started' documentation for guidance.");
-                LOGGER.info("View this and more helpful information here: https://docs.papermc.io/paper/next-steps");
-                LOGGER.info("*************************************************************************************");
-            }
+            // Fiddle start - Disable onboarding message
+            // if (io.papermc.paper.configuration.GlobalConfiguration.isFirstStart) {
+            //     LOGGER.info("*************************************************************************************");
+            //     LOGGER.info("This is the first time you're starting this server.");
+            //     LOGGER.info("It's recommended you read our 'Getting Started' documentation for guidance.");
+            //     LOGGER.info("View this and more helpful information here: https://docs.papermc.io/paper/next-steps");
+            //     LOGGER.info("*************************************************************************************");
+            // }
+            // Fiddle end - Disable onboarding message
             // Paper end - Add onboarding message for initial server start
             // Paper start - Improve outdated version checking
             if (System.getProperty("paper.disableStartupVersionCheck") == null && io.papermc.paper.configuration.GlobalConfiguration.get().updateChecker.enabled) {
