--- a/net/minecraft/server/Services.java
+++ b/net/minecraft/server/Services.java
@@ -24,7 +_,7 @@
 
     // Paper start - add paper configuration files
     public record PaperServices(
-        io.papermc.paper.configuration.PaperConfigurations configurations,
+        io.papermc.paper.configuration.PaperConfigurations.OriginalPaperConfigurations configurations, // Fiddle - Configuration - Use PaperConfigurations as basis - Re-add original concrete class
         io.papermc.paper.profile.PaperFilledProfileCache filledProfileCache
     ) {}
 
@@ -44,8 +_,8 @@
         UserNameToIdResolver userNameToIdResolver = new CachedUserNameToIdResolver(gameProfileRepository, userCacheFile); // Paper
         // Paper start - load paper config files from cli options
         final java.nio.file.Path legacyConfigPath = ((File) optionSet.valueOf("paper-settings")).toPath();
-        final java.nio.file.Path configDirPath = ((File) optionSet.valueOf("paper-settings-directory")).toPath();
-        io.papermc.paper.configuration.PaperConfigurations paperConfigurations = io.papermc.paper.configuration.PaperConfigurations.setup(legacyConfigPath, configDirPath, profileRepository.toPath(), (File) optionSet.valueOf("spigot-settings"));
+        final java.nio.file.Path configDirPath = getConfigDirPath(optionSet); // Fiddle - Configuration - Load global configuration before bootstrap - Extract config dir path getter to method
+        io.papermc.paper.configuration.PaperConfigurations.OriginalPaperConfigurations paperConfigurations = io.papermc.paper.configuration.PaperConfigurations.setup(legacyConfigPath, configDirPath, profileRepository.toPath(), (File) optionSet.valueOf("spigot-settings")); // Fiddle - Configuration - Use PaperConfigurations as basis - Re-add original concrete class
         PaperServices paperServices = new PaperServices(
             paperConfigurations,
             new io.papermc.paper.profile.PaperFilledProfileCache()
@@ -62,4 +_,10 @@
     public boolean canValidateProfileKeys() {
         return !this.servicesKeySet.keys(ServicesKeyType.PROFILE_KEY).isEmpty();
     }
+
+    // Fiddle start - Configuration - Load global configuration before bootstrap - Extract config dir path getter to method
+    public static java.nio.file.Path getConfigDirPath(joptsimple.OptionSet optionSet) {
+        return ((File) optionSet.valueOf("paper-settings-directory")).toPath();
+    }
+    // Fiddle end - Configuration - Load global configuration before bootstrap - Extract config dir path getter to method
 }
