--- a/src/main/java/io/papermc/paper/registry/entry/RegistryEntryMeta.java
+++ b/src/main/java/io/papermc/paper/registry/entry/RegistryEntryMeta.java
@@ -17,6 +_,9 @@
 import org.bukkit.NamespacedKey;
 import org.bukkit.craftbukkit.CraftRegistry;
 import org.bukkit.craftbukkit.util.ApiVersion;
+import org.fiddlemc.fiddle.impl.minecraft.registries.BlockRegistry;
+import org.fiddlemc.fiddle.impl.minecraft.registries.ItemRegistry;
+import org.fiddlemc.fiddle.impl.paper.registry.KeyAwareWritableCraftRegistry;
 
 public sealed interface RegistryEntryMeta<M, A extends Keyed> permits RegistryEntryMeta.ApiOnly, RegistryEntryMeta.ServerSide { // TODO remove Keyed
 
@@ -113,6 +_,11 @@
         @Override
         public org.bukkit.Registry<A> createApiRegistry(final Registry<M> nmsRegistry) {
             if (this.modificationApiSupport.canAdd()) {
+                // Fiddle start - Mode data-driven - Paper API - Make block and item registries key-aware
+                if (nmsRegistry instanceof BlockRegistry || nmsRegistry instanceof ItemRegistry) {
+                    return (org.bukkit.Registry<A>) new KeyAwareWritableCraftRegistry((MappedRegistry<M>) nmsRegistry, this);
+                }
+                // Fiddle end - Mode data-driven - Paper API - Make block and item registries key-aware
                 return new WritableCraftRegistry<>((MappedRegistry<M>) nmsRegistry, this);
             } else {
                 return ServerSide.super.createApiRegistry(nmsRegistry);
