--- a/io/papermc/paper/FeatureHooks.java
+++ b/io/papermc/paper/FeatureHooks.java
@@ -58,9 +_,11 @@
             if (player.connection == null) continue;
 
             final Boolean shouldModify = chunk.getLevel().chunkPacketBlockController.shouldModify(player, chunk);
-            player.connection.send(refreshPackets.computeIfAbsent(shouldModify, s -> { // Use connection to prevent creating firing event
-                return new ClientboundLevelChunkWithLightPacket(chunk, chunk.level.getLightEngine(), null, null, (Boolean) s);
-            }));
+            // Fiddle start - Block mapping - Chunk packets - Pass the player
+            player.connection.send( // Use connection to prevent creating firing event
+                new ClientboundLevelChunkWithLightPacket(chunk, chunk.level.getLightEngine(), null, null, (Boolean) shouldModify, player)
+            );
+            // Fiddle end - Block mapping - Chunk packets - Pass the player
             // Paper end - Anti-Xray
         }
     }
