--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -992,6 +_,7 @@
 
         org.spigotmc.SpigotConfig.init((File) this.console.options.valueOf("spigot-settings")); // Spigot
         this.console.paperConfigurations.reloadConfigs(this.console);
+        org.fiddlemc.fiddle.impl.configuration.FiddleConfigurations.get(null).reloadConfigs(this.console); // Fiddle - Configuration - Reload on Bukkit reload
         for (ServerLevel world : this.console.getAllLevels()) {
             // world.serverLevelData.setDifficulty(config.difficulty); // Paper - per level difficulty
             world.setSpawnSettings(world.isSpawningMonsters()); // Paper - per level difficulty (from MinecraftServer#setDifficulty(ServerLevel, Difficulty, boolean))
@@ -2728,6 +_,14 @@
         public YamlConfiguration getPaperConfig() {
             return CraftServer.this.console.paperConfigurations.createLegacyObject(CraftServer.this.console);
         }
+
+        // Fiddle start - Configuration - API - Implementation
+        @Override
+        public YamlConfiguration getFiddleConfig()
+        {
+            return org.fiddlemc.fiddle.impl.configuration.FiddleConfigurations.get(null).createLegacyObject(CraftServer.this.console);
+        }
+        // Fiddle end - Configuration - API - Implementation
 
         @Override
         public void restart() {
