--- a/src/main/java/org/bukkit/block/BlockType.java
+++ b/src/main/java/org/bukkit/block/BlockType.java
@@ -2744,4 +_,17 @@
      * @return false if this block never has collision, true if it <b>might</b> have collision
      */
     boolean hasCollision();
+
+    // Fiddle start - More data-driven - Vanilla flag - API part
+    /**
+     * Returns whether this block type is present in vanilla.
+     * <p>
+     * This does not take into account whether any details of this block type, such as its hardness,
+     * are different (were modified) from vanilla.
+     * </p>
+     *
+     * @return Whether this block type is present in vanilla.
+     */
+    boolean isVanilla();
+    // Fiddle end - More data-driven - Vanilla flag - API part
 }
