--- a/net/minecraft/world/level/block/Blocks.java
+++ b/net/minecraft/world/level/block/Blocks.java
@@ -7271,6 +_,13 @@
 
     private static Block register(ResourceKey<Block> resourceKey, Function<BlockBehaviour.Properties, Block> factory, BlockBehaviour.Properties properties) {
         Block block = factory.apply(properties.setId(resourceKey));
+        // Fiddle start - Vanilla flag
+        block.markAsVanilla();
+        // Currently, there is no way to modify block states of vanilla blocks, so we assume all block states of vanilla blocks are vanilla too
+        for (BlockState blockState : block.getStateDefinition().getPossibleStates()) {
+            blockState.markAsVanilla();
+        }
+        // Fiddle end - Vanilla flag
         return Registry.register(BuiltInRegistries.BLOCK, resourceKey, block);
     }
 
@@ -7293,9 +_,10 @@
     static {
         for (Block block : BuiltInRegistries.BLOCK) {
             for (BlockState blockState : block.getStateDefinition().getPossibleStates()) {
-                Block.BLOCK_STATE_REGISTRY.add(blockState);
+                // Block.BLOCK_STATE_REGISTRY.add(blockState); // Fiddle - Custom registries for blocks and items - Register block states in BlockRegistry#register
                 blockState.initCache();
             }
         }
+        org.fiddlemc.fiddle.minecraft.registries.BlockRegistry.get().canInitializeBlockStateCaches = true; // Fiddle - Custom registries for blocks and items - Initialize block state caches for non-vanilla blocks in BlockRegistry#register
     }
 }
