--- a/net/minecraft/util/parsing/packrat/Scope.java
+++ b/net/minecraft/util/parsing/packrat/Scope.java
@@ -284,6 +_,12 @@
     private int depth;
     @SuppressWarnings({"unchecked","rawtypes"})
     public static <S> Term<S> increaseDepth() {
+        // Fiddle start - Deobf fix
+        return (Term<S>) IncreasingDepthTerm.INSTANCE;
+    }
+
+    static
+        // Fiddle end - Deobf fix
         class IncreasingDepthTerm<W> implements Term<W> {
             public static final IncreasingDepthTerm INSTANCE = new IncreasingDepthTerm();
             @Override
@@ -295,11 +_,19 @@
                 return true;
             }
         }
-        return (Term<S>) IncreasingDepthTerm.INSTANCE;
-    }
+    // Fiddle start - Deobf fix
+    //     return (Term<S>) IncreasingDepthTerm.INSTANCE;
+    // }
+    // Fiddle end - Deobf fix
 
     @SuppressWarnings({"unchecked","rawtypes"})
     public static <S> Term<S> decreaseDepth() {
+        // Fiddle start - Deobf fix
+        return (Term<S>) DecreasingDepthTerm.INSTANCE;
+    }
+
+    static
+        // Fiddle end - Deobf fix
         class DecreasingDepthTerm<W> implements Term<W> {
             public static final DecreasingDepthTerm INSTANCE = new DecreasingDepthTerm();
             @Override
@@ -308,7 +_,9 @@
                 return true;
             }
         }
-        return (Term<S>) DecreasingDepthTerm.INSTANCE;
-    }
+    // Fiddle start - Deobf fix
+    //     return (Term<S>) DecreasingDepthTerm.INSTANCE;
+    // }
+    // Fiddle end - Deobf fix
     // Paper end - track depth
 }
