--- a/src/main/java/org/bukkit/block/data/BlockData.java
+++ b/src/main/java/org/bukkit/block/data/BlockData.java
@@ -323,4 +_,17 @@
      * @return true if block is replaceable
      */
     boolean isReplaceable();
+
+    // Fiddle start - More data-driven - Vanilla flag - API part
+    /**
+     * Returns whether this block data is present in vanilla.
+     * <p>
+     * This does not take into account whether any details of this block data, such as its collision box,
+     * are different (were modified) from vanilla.
+     * </p>
+     *
+     * @return Whether this block data is present in vanilla.
+     */
+    boolean isVanilla();
+    // Fiddle end - More data-driven - Vanilla flag - API part
 }
