--- a/net/minecraft/network/protocol/game/ClientboundRecipeBookAddPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundRecipeBookAddPacket.java
@@ -11,7 +_,7 @@
 public record ClientboundRecipeBookAddPacket(List<ClientboundRecipeBookAddPacket.Entry> entries, boolean replace) implements Packet<ClientGamePacketListener> {
     public static final StreamCodec<RegistryFriendlyByteBuf, ClientboundRecipeBookAddPacket> STREAM_CODEC = StreamCodec.composite(
         ClientboundRecipeBookAddPacket.Entry.STREAM_CODEC.apply(ByteBufCodecs.list()),
-        ClientboundRecipeBookAddPacket::entries,
+        org.fiddlemc.fiddle.impl.minecraft.packet.mapping.item.recipe.RecipeBookAddPacketEntriesFilterForClientView::getEntriesFilteredForClientView, // Fiddle - Packet mapping - Recipes - Remove recipes that client can't accept
         ByteBufCodecs.BOOL,
         ClientboundRecipeBookAddPacket::replace,
         ClientboundRecipeBookAddPacket::new
