--- a/net/minecraft/network/protocol/game/ClientboundSectionBlocksUpdatePacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundSectionBlocksUpdatePacket.java
@@ -68,7 +_,7 @@
         buffer.writeVarInt(this.positions.length);
 
         for (int i = 0; i < this.positions.length; i++) {
-            buffer.writeVarLong((long)Block.getId(this.states[i]) << 12 | this.positions[i]);
+            buffer.writeVarLong((long)Block.getId(org.fiddlemc.fiddle.impl.packetmapping.block.BlockMappingPipelineImpl.get().apply(this.states[i], new org.fiddlemc.fiddle.impl.packetmapping.block.BlockStateMappingContextImpl(buffer.getClientViewOrFallback(), true))) << 12 | this.positions[i]); // Fiddle - Block mapping - Implementation - Apply during section blocks update write
         }
     }
 
