--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -165,6 +_,12 @@
         return this.paperConfig;
     }
     // Paper end - add paper world config
+    // Fiddle start - Configuration - World configuration
+    private final org.fiddlemc.fiddle.configuration.FiddleWorldConfiguration fiddleConfig;
+    public org.fiddlemc.fiddle.configuration.FiddleWorldConfiguration fiddleConfig() {
+        return this.fiddleConfig;
+    }
+    // Fiddle end - Configuration - World configuration
 
     public static @Nullable BlockPos lastPhysicsProblem; // Spigot
     private int tileTickPosition;
@@ -833,6 +_,8 @@
         org.bukkit.World.Environment environment, // Paper
         java.util.function.Function<org.spigotmc.SpigotWorldConfig, // Spigot - create per world config
         io.papermc.paper.configuration.WorldConfiguration> paperWorldConfigCreator, // Paper - create paper world config
+        java.util.function.Function<org.spigotmc.SpigotWorldConfig, // Spigot - create per world config
+        org.fiddlemc.fiddle.configuration.FiddleWorldConfiguration> fiddleWorldConfigCreator, // Fiddle - Configuration - Create world configuration
         java.util.concurrent.Executor executor // Paper - Anti-Xray
     ) {
         // Paper start - getblock optimisations - cache world height/sections
@@ -846,6 +_,7 @@
         // Paper end - getblock optimisations - cache world height/sections
         this.spigotConfig = new org.spigotmc.SpigotWorldConfig(((net.minecraft.world.level.storage.PrimaryLevelData) levelData).getLevelName()); // Spigot
         this.paperConfig = paperWorldConfigCreator.apply(this.spigotConfig); // Paper - create paper world config
+        this.fiddleConfig = fiddleWorldConfigCreator.apply(this.spigotConfig); // Fiddle - Configuration - Create world configuration
         this.generator = generator;
         this.world = new CraftWorld((ServerLevel) this, generator, biomeProvider, environment);
 
