--- a/src/main/java/io/papermc/paper/antixray/ChunkPacketInfo.java
+++ b/src/main/java/io/papermc/paper/antixray/ChunkPacketInfo.java
@@ -9,7 +_,8 @@
     private final ClientboundLevelChunkWithLightPacket chunkPacket;
     private final LevelChunk chunk;
     private final int[] bits;
-    private final Object[] palettes;
+    public final Object[] palettes; // Fiddle - Block mapping - Chunk packets - Expose palettes
+    public final int[] sectionEndIndexes; // Fiddle - Block mapping - Chunk packets - Track section end indexes
     private final int[] indexes;
     private final Object[][] presetValues;
     private byte[] buffer;
@@ -20,6 +_,7 @@
         int sections = chunk.getSectionsCount();
         bits = new int[sections];
         palettes = new Object[sections];
+        sectionEndIndexes = new int[sections]; // Fiddle - Block mapping - Chunk packets - Track section end indexes
         indexes = new int[sections];
         presetValues = new Object[sections][];
     }
