--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -55,7 +_,15 @@
 import org.slf4j.Marker;
 import org.slf4j.MarkerFactory;
 
-public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
+public class Connection extends SimpleChannelInboundHandler<Packet<?>> implements org.fiddlemc.fiddle.impl.clientview.lookup.ClientViewLookup { // Fiddle - Client view - Lookup - Implement for Connection
+    private org.fiddlemc.fiddle.api.clientview.@Nullable ClientView clientView; // Fiddle - Client view - Set - Field
+    // Fiddle start - Client view - Lookup - Implement for Connection - Getter
+    @Override
+    public org.fiddlemc.fiddle.api.clientview.@Nullable ClientView getClientView() {
+        return this.clientView;
+    }
+    // Fiddle end - Client view - Lookup - Implement for Connection - Getter
+    public final org.fiddlemc.fiddle.impl.packetmapping.item.reverse.ItemMappingReverser itemMappingReverser; // Fiddle - Item mapping - Reverse - Store reverser in Connection
     private static final float AVERAGE_PACKETS_SMOOTHING = 0.75F;
     private static final Logger LOGGER = LogUtils.getLogger();
     public static final Marker ROOT_MARKER = MarkerFactory.getMarker("NETWORK");
@@ -130,12 +_,15 @@
 
     public Connection(PacketFlow receiving) {
         this.receiving = receiving;
+        this.clientView = new org.fiddlemc.fiddle.impl.clientview.JavaDefaultClientViewImpl(this)/* TODO determine actual view */; // Fiddle - Client view - Set - Field
+        this.itemMappingReverser = new org.fiddlemc.fiddle.impl.packetmapping.item.reverse.ItemMappingReverser(); // Fiddle - Item mapping - Reverse - Store reverser in Connection
     }
 
     @Override
     public void channelActive(ChannelHandlerContext ctx) throws Exception {
         super.channelActive(ctx);
         this.channel = ctx.channel();
+        this.channel.attr(org.fiddlemc.fiddle.impl.clientview.lookup.packethandling.NettyClientViewLookupAttribute.ATTRIBUTE_KEY).set(new java.lang.ref.WeakReference<>(this)); // Fiddle - Client view - Lookup - During packet handling - Copy the connection (acting as the lookup) to the channel
         this.address = this.channel.remoteAddress();
         this.preparing = false; // Spigot
         if (this.delayedDisconnect != null) {
