--- a/net/minecraft/network/FriendlyByteBuf.java
+++ b/net/minecraft/network/FriendlyByteBuf.java
@@ -68,9 +_,17 @@
 import org.joml.Vector3fc;
 import org.jspecify.annotations.Nullable;
 
-public class FriendlyByteBuf extends ByteBuf {
+public class FriendlyByteBuf extends ByteBuf implements org.fiddlemc.fiddle.client.ClientViewProvider { // Fiddle - Client view - Store provider in FriendlyByteBuf - Easy access via ClientViewProvider interface
     private final ByteBuf source;
     public final java.util.@Nullable Locale adventure$locale; // Paper - track player's locale for server-side translations
+    public final java.lang.ref.@Nullable WeakReference<org.fiddlemc.fiddle.client.ClientViewProvider> clientViewProvider; // Fiddle - Client view - Store provider in FriendlyByteBuf
+    // Fiddle start - Client view - Store provider in FriendlyByteBuf - Easy access via ClientViewProvider interface
+    @Override
+    public org.fiddlemc.fiddle.client.@Nullable ClientView getClientView() {
+        org.fiddlemc.fiddle.client.@Nullable ClientViewProvider clientViewProvider = this.clientViewProvider == null ? null : this.clientViewProvider.get();
+        return clientViewProvider == null ? null : clientViewProvider.getClientView();
+    }
+    // Fiddle - Client view - Store provider in FriendlyByteBuf - Easy access via ClientViewProvider interface
     public static final short MAX_STRING_LENGTH = 32767;
     public static final int MAX_COMPONENT_STRING_LENGTH = 262144;
     private static final int PUBLIC_KEY_SIZE = 256;
@@ -84,6 +_,7 @@
 
     public FriendlyByteBuf(ByteBuf source) {
         this.adventure$locale = PacketEncoder.ADVENTURE_LOCALE.get(); // Paper - track player's locale for server-side translations
+        this.clientViewProvider = org.fiddlemc.fiddle.client.ClientViewProviderThreadLocal.THREAD_LOCAL.get(); // Fiddle - Client view - Store provider in FriendlyByteBuf
         this.source = source;
     }
 
