--- a/net/minecraft/network/HashedStack.java
+++ b/net/minecraft/network/HashedStack.java
@@ -21,7 +_,24 @@
             return stack.isEmpty();
         }
     };
-    StreamCodec<RegistryFriendlyByteBuf, HashedStack> STREAM_CODEC = ByteBufCodecs.optional(HashedStack.ActualItem.STREAM_CODEC)
+    // Fiddle start - Item mapping - Reverse - Reverse-map hashed item stacks - Wrap codec
+    StreamCodec<RegistryFriendlyByteBuf, HashedStack> STREAM_CODEC = new net.minecraft.network.codec.StreamCodec<RegistryFriendlyByteBuf, HashedStack>() {
+
+        @Override
+        public HashedStack decode(final RegistryFriendlyByteBuf buffer) {
+            HashedStack decoded = UNMAPPED_STREAM_CODEC.decode(buffer);
+            org.fiddlemc.fiddle.impl.packetmapping.item.reverse.@org.jspecify.annotations.Nullable ItemMappingReverser itemMappingReverser = ((org.fiddlemc.fiddle.impl.clientview.ClientViewImpl) buffer.getClientViewOrFallback()).getItemMappingReverser();
+            return itemMappingReverser != null ? itemMappingReverser.reverseMappingIfPossible(decoded) : decoded;
+        }
+
+        @Override
+        public void encode(final RegistryFriendlyByteBuf buffer, final HashedStack value) {
+            UNMAPPED_STREAM_CODEC.encode(buffer, value);
+        }
+
+    };
+    StreamCodec<RegistryFriendlyByteBuf, HashedStack> UNMAPPED_STREAM_CODEC = ByteBufCodecs.optional(HashedStack.ActualItem.STREAM_CODEC)
+        // Fiddle end - Item mapping - Reverse - Reverse-map hashed item stacks - Wrap codec
         .map(
             optional -> DataFixUtils.orElse((Optional<? extends HashedStack>)optional, EMPTY),
             hashedStack -> hashedStack instanceof HashedStack.ActualItem actualItem ? Optional.of(actualItem) : Optional.empty()
