--- a/net/minecraft/network/PacketListener.java
+++ b/net/minecraft/network/PacketListener.java
@@ -8,7 +_,7 @@
 import net.minecraft.network.protocol.PacketFlow;
 import net.minecraft.network.protocol.PacketUtils;
 
-public interface PacketListener {
+public interface PacketListener extends org.fiddlemc.fiddle.impl.client.view.lookup.ClientViewLookup { // Fiddle - Client view lookup - Implement for PacketListener (actually only implemented for ServerLoginPacketListenerImpl and ServerCommonPacketListenerImpl, but having the method available is convenient)
     PacketFlow flow();
 
     ConnectionProtocol protocol();
