From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Sun, 18 Feb 2024 15:44:18 +0100
Subject: [PATCH] Temporary fix for Class.getEnumConstants issue with Material


diff --git a/src/main/java/org/bukkit/Registry.java b/src/main/java/org/bukkit/Registry.java
index e1fb4d8cca6a9c59047b1396f5c40bea957d777a..98bd2501a5c932b30e1fb9573a372c20d9717ac3 100644
--- a/src/main/java/org/bukkit/Registry.java
+++ b/src/main/java/org/bukkit/Registry.java
@@ -396,7 +396,7 @@ public interface Registry<T extends Keyed> extends Iterable<T> {
         protected SimpleRegistry(@NotNull Class<T> type, @NotNull Predicate<T> predicate) {
             ImmutableMap.Builder<NamespacedKey, T> builder = ImmutableMap.builder();
 
-            for (T entry : type.getEnumConstants()) {
+            for (T entry : type.equals(org.bukkit.Material.class) ? ((T[]) org.bukkit.Material.values()) : type.getEnumConstants()) { // Fiddle - modifiable Bukkit enums - inject runtime versions - Material
                 if (predicate.test(entry)) {
                     builder.put(entry.getKey(), entry);
                 }
