From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Sat, 30 Dec 2023 17:04:19 +0100
Subject: [PATCH] temp comment out compute modified buffer for chunk data


diff --git a/src/main/java/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java b/src/main/java/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
index 278d9a9a1eb367042b52a12581c4f4f9736990c8..8fffc832f5de5d068f8e09b28a2d8e71bfbb177d 100644
--- a/src/main/java/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
+++ b/src/main/java/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
@@ -204,7 +204,7 @@ public class ClientboundLevelChunkPacketData {
 
     private void startModifyBufferTask(@NotNull ClientPerspective clientPerspective, @NotNull ModifiedClientboundLevelChunkPacketDataBuffer modifiedBuffer) {
         // Run the task inlined on the current thread
-        BlockStateReplacer.computeModifiedBufferForChunkData(this, clientPerspective, modifiedBuffer);
+//        BlockStateReplacer.computeModifiedBufferForChunkData(this, clientPerspective, modifiedBuffer);
     }
     // Fiddle end - client perspective - block visualization - modify chunk packets
 
