From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Mon, 29 Jan 2024 16:34:28 +0100
Subject: [PATCH] fixup build.gradle.kts Fiddle changes: project setup


diff --git a/build.gradle.kts b/build.gradle.kts
index 8182ab132f82502554739562ba7c2e5e429e68bc..7ce6a5a0788d949eb9355ded4561c42f05ee62a6 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -18,8 +18,12 @@ dependencies {
     // Fiddle start - project setup
     // Depend on own API
     implementation(project(":$apiProjectName"))
+    // Depend on Paper API
+    implementation("io.papermc.paper:paper-api:1.20.4-R0.1-SNAPSHOT") {
+        exclude("io.papermc.paper", "paper-api")
+    }
     // Depend on Paper MojangAPI
-    implementation("io.papermc.paper:paper-mojangapi:1.20.1-R0.1-SNAPSHOT") {
+    implementation("io.papermc.paper:paper-mojangapi:1.20.4-R0.1-SNAPSHOT") {
         exclude("io.papermc.paper", "paper-api")
     }
     // Fiddle end - project setup
