From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Mon, 12 Feb 2024 20:53:25 +0100
Subject: [PATCH] Temp data converter fix until update Paper upstream


diff --git a/src/main/java/ca/spottedleaf/dataconverter/minecraft/versions/V3568.java b/src/main/java/ca/spottedleaf/dataconverter/minecraft/versions/V3568.java
index 98272d556a09e9f035005c6783ebb229c2c7157d..d350bf9dab8353932a3d1ee8d2117f785506016e 100644
--- a/src/main/java/ca/spottedleaf/dataconverter/minecraft/versions/V3568.java
+++ b/src/main/java/ca/spottedleaf/dataconverter/minecraft/versions/V3568.java
@@ -201,6 +201,12 @@ public final class V3568 {
 
                 final MapType<String> tag = root.getMap("tag");
 
+                // Fiddle start - temp fix until update Paper upstream
+                if (tag == null) {
+                    return null;
+                }
+                // Fiddle end - temp fix until update Paper upstream
+
                 if ("minecraft:suspicious_stew".equals(id)) {
                     RenameHelper.renameSingle(tag, "Effects", "effects");
 
