--- a/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java
@@ -797,4 +_,11 @@
     public boolean isReplaceable() {
         return this.state.canBeReplaced();
     }
+
+    // Fiddle start - More data-driven - Vanilla flag
+    @Override
+    public boolean isVanilla() {
+        return this.state.isVanilla();
+    }
+    // Fiddle end - More data-driven - Vanilla flag
 }
