--- a/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java
@@ -603,7 +_,7 @@
                 }
 
                 StringReader reader = new StringReader(data);
-                BlockStateParser.BlockResult arg = BlockStateParser.parseForBlock(CraftRegistry.getMinecraftRegistry(Registries.BLOCK), reader, false);
+                BlockStateParser.BlockResult arg = BlockStateParser.parseForBlock(BuiltInRegistries.BLOCK/*CraftRegistry.getMinecraftRegistry(Registries.BLOCK)*/, reader, false); // Fiddle - More data-driven - Minecraft registries - Use registry directly instead of looking it up by key, to allow translating Bukkit to Minecraft block state classes before the block registry is registered with the registry of registries
                 Preconditions.checkArgument(!reader.canRead(), "Spurious trailing data: " + data);
 
                 blockData = arg.blockState();
@@ -797,4 +_,11 @@
     public boolean isReplaceable() {
         return this.state.canBeReplaced();
     }
+
+    // Fiddle start - More data-driven - Vanilla flag - Implementation part
+    @Override
+    public boolean isVanilla() {
+        return this.state.isVanilla();
+    }
+    // Fiddle end - More data-driven - Vanilla flag - Implementation part
 }
