--- a/src/main/java/org/bukkit/inventory/ItemType.java
+++ b/src/main/java/org/bukkit/inventory/ItemType.java
@@ -3345,4 +_,17 @@
      */
     @ApiStatus.Experimental
     @Unmodifiable Set<DataComponentType> getDefaultDataTypes();
+
+    // Fiddle start - Vanilla flag - API
+    /**
+     * Returns whether this item type is present in vanilla.
+     * <p>
+     * This does not take into account whether any details of this item type, such as its max stack size,
+     * are different (were modified) from vanilla.
+     * </p>
+     *
+     * @return Whether this item type is present in vanilla.
+     */
+    boolean isVanilla();
+    // Fiddle end - Vanilla flag - API
 }
