--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -261,6 +_,7 @@
         this.paperConfigurations.initializeGlobalConfiguration(this.registryAccess());
         this.paperConfigurations.initializeWorldDefaultsConfiguration(this.registryAccess());
         // Paper end - initialize global and world-defaults configuration
+        org.fiddlemc.fiddle.configuration.FiddleConfigurations.get(this.options).initializeWorldDefaultsConfiguration(this.registryAccess()); // Fiddle - Configuration - Initialize world-defaults configuration
         this.server.spark.enableEarlyIfRequested(); // Paper - spark
         // Paper start - fix converting txt to json file; convert old users earlier after PlayerList creation but before file load/save
         if (this.convertOldUsers()) {
@@ -315,6 +_,20 @@
             if (true) throw new IllegalStateException("Failed to bind to port", var11); // Paper - Propagate failed to bind to port error
             return false;
         }
+
+        // Fiddle start - Synchronize Bukkit enums - Run on server start
+        try {
+            @Nullable String errorMessage = org.fiddlemc.fiddle.bukkit.enums.MaterialEnumSynchronizer.run();
+            if (errorMessage != null) {
+                // Don't start the server with an incomplete Material enum
+                LOGGER.warn("Failed to extend Bukkit Material enum: {}", errorMessage);
+                return false;
+            }
+        } catch (Exception e) {
+            // Don't start the server with an incomplete Material enum
+            throw new IllegalStateException("Failed to extend Bukkit Material enum", e);
+        }
+        // Fiddle end - Synchronize Bukkit enums - Run on server start
 
         // CraftBukkit start
         this.server.loadPlugins();
