--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -261,6 +_,7 @@
         this.paperConfigurations.initializeGlobalConfiguration(this.registryAccess());
         this.paperConfigurations.initializeWorldDefaultsConfiguration(this.registryAccess());
         // Paper end - initialize global and world-defaults configuration
+        org.fiddlemc.fiddle.impl.configuration.FiddleConfigurations.get(this.options).initializeWorldDefaultsConfiguration(this.registryAccess()); // Fiddle - Configuration - Initialize world-defaults configuration
         this.server.spark.enableEarlyIfRequested(); // Paper - spark
         // Paper start - fix converting txt to json file; convert old users earlier after PlayerList creation but before file load/save
         if (this.convertOldUsers()) {
@@ -275,6 +_,12 @@
         com.destroystokyo.paper.Metrics.PaperMetrics.startMetrics(); // Paper - start metrics
         com.destroystokyo.paper.VersionHistoryManager.INSTANCE.getClass(); // Paper - load version history now
 
+        // Fiddle start - Packet mapping - Register replacements
+        org.fiddlemc.fiddle.impl.minecraft.packet.mapping.item.ItemMappingPipelineImpl.ComposeEventImpl itemMappingPipelienComposeEvent = new org.fiddlemc.fiddle.impl.minecraft.packet.mapping.item.ItemMappingPipelineImpl.ComposeEventImpl();
+        io.papermc.paper.plugin.lifecycle.event.LifecycleEventRunner.INSTANCE.callEvent(org.fiddlemc.fiddle.impl.minecraft.packet.mapping.item.ItemMappingPipelineImpl.get().compose(), itemMappingPipelienComposeEvent);
+        itemMappingPipelienComposeEvent.bake();
+        // Fiddle end - Packet mapping - Register replacements
+
         // this.worldData.setGameType(properties.gameMode.get()); // CraftBukkit - moved to world loading
         LOGGER.info("Default game type: {}", properties.gameMode.get());
         // Paper start - Unix domain socket support
@@ -315,6 +_,20 @@
             if (true) throw new IllegalStateException("Failed to bind to port", var11); // Paper - Propagate failed to bind to port error
             return false;
         }
+
+        // Fiddle start - Synchronize Bukkit enums - Run on server start
+        try {
+            @Nullable String errorMessage = org.fiddlemc.fiddle.impl.bukkit.enums.MaterialEnumSynchronizer.run();
+            if (errorMessage != null) {
+                // Don't start the server with an incomplete Material enum
+                LOGGER.warn("Failed to extend Bukkit Material enum: {}", errorMessage);
+                return false;
+            }
+        } catch (Exception e) {
+            // Don't start the server with an incomplete Material enum
+            throw new IllegalStateException("Failed to extend Bukkit Material enum", e);
+        }
+        // Fiddle end - Synchronize Bukkit enums - Run on server start
 
         // CraftBukkit start
         this.server.loadPlugins();
